### ISSUE TITLE
XFA - Support tiff images

### ISSUE THREAD

**Issue by calixteman** [mozilla/pdf.js#13707]
Firefox doesn't support natively tiff format so we need to implement a decoder for XFA.
As far as I know there are not so much pdf containing some tiff images.

**Comment by calixteman** [comment]
Maybe we could use https://github.com/image-rs/image-tiff and compile it into wasm.

**Comment by Snuffleupagus** [comment]
Assuming it'd work for the sort of TIFF images found in XFA documents, wouldn't a JavaScript solution be *a lot* simpler since it could be included directly using e.g. NPM; I've not tested it, but a quick web-search turned up https://www.npmjs.com/package/tiff

**Comment by calixteman** [comment]
@Snuffleupagus, would you like to fix this bug ?

**Comment by CoolCanuck** [comment]
Attached are some PDFs that use TIF or TIFF (top left logo), as it may be helpful
[4808-69E (1).pdf](https://github.com/mozilla/pdf.js/files/7190869/4808-69E.1.pdf   - **top** left image
[c0009_re_1587386188899_eng.pdf](https://github.com/mozilla/pdf.js/files/7190877/c0009_re_1587386188899_eng.pdf) - **bottom** right image

**Comment by khairnarsaurabh23** [comment]
Can i get this bug to get fixed?

**Comment by timvandermeij** [comment]
Sure, feel free to work on this.

**Comment by abhpd** [comment]
Hey @calixteman As suggested by @Snuffleupagus, I was trying to import [tiff](https://www.npmjs.com/package/tiff) to `src/core/xfa/template.js`. However direct import doesn't seem to work, can you guide me on how we are handling imports in the project?

<img width="732" alt="Screenshot 2021-10-06 at 10 32 51 PM" src="https://user-images.githubusercontent.com/34980034/136250327-31d289aa-1443-499e-8a14-a81ad003a633.png">


Here is the error:
<img width="643" alt="Screenshot 2021-10-06 at 10 29 47 PM" src="https://user-images.githubusercontent.com/34980034/136249839-ad61f849-ffb7-46bc-a6a5-5b42c0f1801e.png">

Also What I understand is that we need to decode the `buffer` using tiff or is it something else?
For your reference:
<img width="915" alt="Screenshot 2021-10-06 at 10 28 07 PM" src="https://user-images.githubusercontent.com/34980034/136249942-a6f28ab1-ce1f-429d-bd57-44213169226d.png">

**Comment by abhpd** [comment]
@OladapoAjala Yes, I've posted an issue I was facing above your comment, let me know if you have any idea.

**Comment by brendandahl** [comment]
Instead of doing npm install, I think you'll want to:
1) build a fully bundled version of tiff.js (hopefully this is possible, I haven't looked into it)
2) make a new `tiff` directory in [external](https://github.com/mozilla/pdf.js/tree/master/external)
3) move tiff.js and it's license in there
4) import it from there e.g. `import "../../external/tiff/tiff.js`

Once you have that working we should work on lazy loading the tiff module since it won't be used very often. This will require more work and bundling another file with the pdf.js build.

**Comment by alinethome** [comment]
Is anyone still working on this? I would love to work on it if not. Or collaborate if someone is. 

I did some tinkering and managed to figure out why the imports weren't working (the TIFF library was importing node's encoding and decoding utility, which webpack no longer has a default polyfill for; simply using the browser's API or a suitable polyfill should fix the problem).

**Comment by calixteman** [comment]
@alinethome, no need to ask, just work on it.
You can submit a PR (even in draft mode) and we can then iterate on it.
What's the tiff lib you want to use ? its size ?

**Comment by alinethome** [comment]
@calixteman Alright, good to know! 

I was going to try with https://www.npmjs.com/package/tiff, Bundling it as @brendandahl suggested yielded a 255 kb file.

**Comment by Snuffleupagus** [comment]
> I was going to try with https://www.npmjs.com/package/tiff, Bundling it as @brendandahl suggested yielded a 255 kb file.

That sounds like *a lot* for a feature that's mostly unused (most XFA doesn't seem to include TIFF images), since it'd correspond to an approximately 14 percent *increase* of the `pdf.worker.js` file size in the MOZCENTRAL build.

---

> Maybe we could use https://github.com/image-rs/image-tiff and compile it into wasm.

Given the size reported above, I cannot help wondering if this idea from https://github.com/mozilla/pdf.js/issues/13707#issuecomment-877378361 would actually be necessary!?

**Comment by alinethome** [comment]
@Snuffleupagus I originally thought I would try lazy loading once I got the basic logic to work, but  I can definitely try to do that instead. The file size is without minifying, but I suppose it would still be quite big once minified.

**Comment by Snuffleupagus** [comment]
> I originally thought I would try lazy loading once I got the basic logic to work, but I can definitely try to do that instead.

Even if the TIFF image decoder is loaded lazily, the file itself still needs to be shipped together with the PDF.js library (in e.g. Mozilla Firefox).

> The file size is without minifying, but I suppose it would still be quite big once minified.

Indeed, even when minified it'd risk being fairly large.

Also, to put the quoted TIFF decoder file size (i.e. 255 kB) into perspective, please note that the PDF.js library currently include image decoders for the JPEG, JPEG 2000, and JBIG2 image formats and those three *together* are actually smaller than that.

Hence why it'd be *interesting* to compare the size of the Rust library compiled to Wasm, as was suggested earlier in the thread.

**Comment by calixteman** [comment]
I added 
```
[profile.release]
opt-level = 's'
```
in the cargo file to optimize on size
and without jpeg and flate2, I get a size around 163K.

**Comment by ranadip01in** [comment]
can i work in this issue, if this issue isn't solve yet

**Comment by Snuffleupagus** [comment]
*Given the discussion above it's not clear to me that this is actually a "good beginner bug", hence removing the label.*
